src/store/index-level.ts

```diff
@@ -1,213 +1,393 @@
-import type { Filter, RangeFilter } from '../types/message-types.js';
-import type { LevelWrapperBatchOperation, LevelWrapperIteratorOptions } from './level-wrapper.js';
+import type { EqualFilter, Filter, KeyValues, QueryOptions, RangeFilter } from '../types/query-types.js';
+import type { LevelWrapperBatchOperation, LevelWrapperIteratorOptions, } from './level-wrapper.js';
 
-import { executeUnlessAborted } from '../utils/abort.js';
-import { flatten } from '../utils/object.js';
+import { lexicographicalCompare } from '../utils/string.js';
+import { SortDirection } from '../types/query-types.js';
 import { createLevelDatabase, LevelWrapper } from './level-wrapper.js';
+import { DwnError, DwnErrorCode } from '../core/dwn-error.js';
+import { FilterSelector, FilterUtility } from '../utils/filter.js';
+import { isEmptyObject, removeEmptyStrings } from '../utils/object.js';
+
+type IndexLevelConfig = {
+  location?: string,
+  createLevelDatabase?: typeof createLevelDatabase
+};
+
+type IndexedItem = { itemId: string, indexes: KeyValues };
+
+const INDEX_SUBLEVEL_NAME = 'index';
 
 export interface IndexLevelOptions {
   signal?: AbortSignal;
 }
 
 /**
- * A LevelDB implementation for indexing the messages stored in the DWN.
+ * A LevelDB implementation for indexing the messages and events stored in the DWN.
  */
 export class IndexLevel {
-  config: IndexLevelConfig;
-
   db: LevelWrapper<string>;
+  config: IndexLevelConfig;
 
   constructor(config: IndexLevelConfig) {
     this.config = {
       createLevelDatabase,
-      ...config
+      ...config,
     };
 
-    this.db = new LevelWrapper<string>({ ...this.config, valueEncoding: 'utf8' });
+    this.db = new LevelWrapper<string>({
+      location            : this.config.location!,
+      createLevelDatabase : this.config.createLevelDatabase,
+      keyEncoding         : 'utf8'
+    });
   }
 
   async open(): Promise<void> {
-    return this.db.open();
+    await this.db.open();
   }
 
   async close(): Promise<void> {
-    return this.db.close();
+    await this.db.close();
+  }
+
+  /**
+ * deletes everything in the underlying index db.
+ */
```

Comment:
Nit: incorrect indentation, might be worthwhile checking if there is away to just have linter detect/fix this.